DJAN-107 tequila adoption

jbradberry opened the pull request:
Merge together the `worker` and `app` roles.  Move off into its own repo, to support installation using `ansible-galaxy`.

jbradberry committed:
Moved the creation of the project database and database user to db-master.

These are tasks that are tied to the database service, regardless of
whether the db is local or on another server.

jbradberry committed:
Improved the installation of rabbitmq.

- add the erlang ppa when the Ubuntu release is 'precise'
- get the latest version of rabbitmq and erlang-nox

jbradberry committed:
Merged the cache role into web.

jbradberry committed:
Merged the web role into the balancer role.

jbradberry committed:
Merged structure role into common.

jbradberry committed:
Removed the empty worker role.

jbradberry committed:
Removed the remainder of the role metadata.

jbradberry committed:
Lifted the app role out to the top level.

And removed all other roles and unnecessary files.

jbradberry committed:
Added ansible-galaxy metadata.

jbradberry committed:
Added tasks to install and set up gunicorn.

But only as a single server, not behind a load balancer.

jbradberry committed:
Added tasks to install and manage celery and celery beat.

jbradberry committed:
Added a task to install newrelic when the use_newrelic variable is set.

jbradberry committed:
Added tasks to install nodejs and npm packages.

dpoirier commented:
This might be a good time to just switch to using the system `supervisor` package.

dpoirier commented:
It seems like we should move `vars/main.yml` to `defaults/main.yml` so projects can override the default values, and then add defaults for all the variables we refer to in this role somewhere.

jbradberry commented:
Ah, right.  Yes to both of those.

dpoirier commented on tasks/main.yml:
```diff
@@ -157,6 +165,80 @@
+- name: symlink less
+  file: src=/usr/bin/lessc
+        dest=/usr/local/bin/lessc
+        state=link
```
This symlink surprises me. If we have `lessc` in `/usr/bin`, why would we need a symlink in `/usr/local/bin`?

jbradberry replied:
@dpoirier because of gunicorn (https://github.com/caktus/margarita/issues/75) it seems.

dpoirier replied:
Ansible confused me - they have "src" and "dest" mixed up on symlinks from how I think of them.  

I think of "ln -s" as analogous to "cp", so "cp src dest" creates a copy of src and puts it at dest, and "ln -s src dest" creates a symlink to src and puts it at dest.  

But Ansible is using "src" and "dest" exactly opposite to that here - this state creates a symlink to dest and puts it at src.

jbradberry replied:
No, looks like I screwed it up.  From the docs:

```
- file: src=/file/to/link/to dest=/path/to/symlink owner=foo group=foo state=link
```

`/usr/local/bin/lessc` is the npm installed path, so it needs to be the src.

dpoirier commented:
Maybe also delete .gitignore from here, and update the year in the LICENSE file, and remove irrelevant content from the README?

This README will need more work than some of the other roles. We make a lot of assumptions about the project.

jbradberry committed:
Added an explanitory comment to the 'symlink less' task.

It's because gunicorn only looks in /usr/local/bin by default, apparently.

jbradberry committed:
Fixed the src and dest for symlinking lessc where gunicorn can find it.

jbradberry committed:
Minor tweaks.

* removed the no longer needed .gitignore file
* updated the year in LICENSE

jbradberry committed:
Moved vars/main.yml to defaults/main.yml.

jbradberry committed:
Updated the README and the defaults.

* removed the project structure information
* specified how to install the role using ansible-galaxy
* listed all of the variables actually used by this role

dpoirier commented:
Can we default 'source_is_local' to false?

jbradberry committed:
Added a default to false for source_is_local.

dpoirier commented on templates/celery.conf:
```diff
@@ -0,0 +1,18 @@
+[program:{{ project_name }}-celery-{{ item.name }}]
+command={{ source_dir }}/dotenv.sh {% if new_relic_license_key %}{{ venv_dir }}/bin/newrelic-admin run-program {% endif %}{{ venv_dir }}/bin/celery -A {{ project_name }} {{ item.command }} {{ item.flags }}
```
This fails if new_relic_license_key is undefined. Let's check use_newrelic instead, here and in gunicorn.conf.

dpoirier commented on tasks/main.yml:
```diff
@@ -133,6 +133,14 @@
        virtualenv_python=/usr/bin/python{{ python_version }}
```
Two lines up from here: the way we're doing this only works if the requirements file is in a subdirectory called requirements.  Not all projects do that: I just tested with https://github.com/caktus/taytay, for example, which just put its requirements files in its top-level source dir.  Let's let people specify the full path into the source tree.

dpoirier replied:
Also, the requirements seems to be a dictionary keyed by the environment. Let's be consistent with other variables and just make this a variable, and the project can set it in the environment's variables if they want, or globally.

dpoirier replied:
Finally (I hope!), let's not assume requirements files end in `.txt`.

dpoirier commented:
We should only run our tasks to set up celery if the host being configured is a worker, and only run the gunicorn ones if the host being configured is a web server.

dpoirier commented on tasks/main.yml:
```diff
@@ -157,37 +165,98 @@
+    - restart celery-beat
+
+- name: ensure celery is running
+  supervisorctl: name={{ project_name }}-celery-default state=started
```
This is failing: FAILED! => {"changed": false, "failed": true, "msg": "ERROR (no such process)", "name": "our_project-celery-default"}

The conf file on the server does have `[program:our_project-celery-default]`. I'm guessing that we need to make supervisor reload its config files before trying this, because at this point it doesn't know we've added this process yet.  Or maybe we should do this in a handler or something like that?

dpoirier replied:
Well, I manually reloaded supervisor and ran the deploy again and got a different error: `TASK [tequila-django : ensure celery is running] *******************************
fatal: [server1]: FAILED! => {"changed": false, "failed": true, "msg": "our_project-celery-default: ERROR (no such file)\n"}
`

I'm pretty sure that's just because this project doesn't use celery and so doesn't install it. But it does suggest that (if we are using celery) that reloading supervisor before trying this task will get past the previous error.

dpoirier commented on tasks/main.yml:
```diff
@@ -157,37 +165,98 @@
+- name: add nodejs repo
+  apt_repository: repo="deb https://deb.nodesource.com/node_4.x {{ ansible_distribution_release }} main"
+                  state=present
+                  filename='nodesource'
```
Error: `TASK [tequila-django : add nodejs repo] ****************************************
fatal: [server1]: FAILED! => {"changed": false, "failed": true, "msg": "unsupported parameter for module: filename"}
`

jbradberry committed:
Removed the 'filename' parameter from the task adding the nodejs repo.

'filename' isn't available until Ansible 2.1.

jbradberry committed:
Promoted the celery restart handlers to tasks.

So that the "ensure celery is running" task can run after them.

jbradberry committed:
Made web and worker tasks conditional on variables set in the project.

jbradberry committed:
Made the path to the requirements file a variable, with a default.

jbradberry committed:
Made the config templates check `use_newrelic` instead of the license key.

Since it has a default value, and the key does not.

dpoirier commented:
Suggest adding to "checkout latest project source": `ssh_opts="-o StrictHostKeyChecking=no"`.

dpoirier commented:
`pip install` should first chdir to the source directory so we can specify the requirements file as a relative path.

dpoirier commented:
I suggest this change to fix the install for nodejs:

```
diff --git a/tasks/web.yml b/tasks/web.yml
index d97ce0c..91b061f 100644
--- a/tasks/web.yml
+++ b/tasks/web.yml
@@ -1,11 +1,12 @@
 ---
-- name: add nodejs repo
-  apt_repository: repo="deb https://deb.nodesource.com/node_4.x {{ ansible_distribution_release }} main"
-                  state=present
-
 - name: add nodejs repo key
   apt_key: data="{{ lookup('file', 'nodesource.pub') }}" state=present

+- name: add nodejs repo
+  apt_repository: repo="deb https://deb.nodesource.com/node_4.x {{ ansible_distribution_release }} main"
+                  state=present update_cache=yes
+  # Note: the update_cache above should only run if we changed the system's repository config
+
 - name: ensure that nodejs-legacy is not installed
   apt: pkg=nodejs-legacy state=absent
```

dpoirier commented:
Better:

```
diff --git a/tasks/web.yml b/tasks/web.yml
index d97ce0c..616f2a8 100644
--- a/tasks/web.yml
+++ b/tasks/web.yml
@@ -1,26 +1,38 @@
 ---
-- name: add nodejs repo
-  apt_repository: repo="deb https://deb.nodesource.com/node_4.x {{ ansible_distribution_release }} main"
-                  state=present
+- name: see if we have package.json
+  command: /bin/test -e {{ source_dir }}/package.json
+  register: has_package_json
+  ignore_errors: true

 - name: add nodejs repo key
   apt_key: data="{{ lookup('file', 'nodesource.pub') }}" state=present
+  when: has_package_json|success
+
+- name: add nodejs repo
+  apt_repository: repo="deb https://deb.nodesource.com/node_4.x {{ ansible_distribution_release }} main"
+                  state=present update_cache=yes
+  # Note: the update_cache above should only run if we changed the system's repository config
+  when: has_package_json|success

 - name: ensure that nodejs-legacy is not installed
   apt: pkg=nodejs-legacy state=absent
+  when: has_package_json|success

 - name: install nodejs
   apt: pkg=nodejs state=latest
+  when: has_package_json|success

 - name: npm install packages
   npm: path={{ source_dir }} state=latest
   become_user: "{{ project_user }}"
+  when: has_package_json|success

 - name: npm run build
   command: npm run build
   args:
     chdir: "{{ source_dir }}"
   become_user: "{{ project_user }}"
+  when: has_package_json|success

 # TODO: find out if the latest gunicorn is ok, or if we really do want
 # to pin to a version in the 19.1 series.
```

dpoirier commented:
Oops - change that last one so that we always install npm & node.  We'll need them farther down to install less.

jbradberry committed:
Reverse the order of operations for adding the nodejs key and repo

jbradberry committed:
Add the requirements_extra_args optional variable

Needed for things like --find-links.

jbradberry committed:
Update the README.rst to include the new recommended roles path

jbradberry committed:
Make npm tasks run only when a package.json file exists

jbradberry committed:
Check for the build script before running the optional npm.yml playbook

jbradberry committed:
Add task to upload the dotenv.sh helper script

jbradberry committed:
Make sure to reread the supervisord conf files

jbradberry committed:
Add task to create project.pth

adding the project source directory to the python path

jbradberry committed:
Fix the whitespace around the optional environment variables

jbradberry committed:
Immedately reread and reload supervisord on configuring gunicorn

jbradberry committed:
Add an optional syncdb task

jbradberry committed:
Clear out pip build cruft

Specialty Food, for example, builds from sdist.  If a previous run had
left these in place, pip installing the requirements will fail on
subsequent runs.

jbradberry committed:
Add the ability to override the celery worker arguments

see also margarita#143

jbradberry committed:
Remove the separate step to install less

It adds a separate unneeded package if the project is using e.g. sass
instead.  Also, deal with the path to less issue (margarita#75) by
adding /usr/local/bin to the PATH variable in the .env file.

jbradberry committed:
Set PYTHONUNBUFFERED=1 to allow stdout to immediately go to the logs

jbradberry committed:
Alter the nodejs/npm build workflow

Skip it entirely if package.json doesn't exist, and only run `npm run
build` when a build script exists.

jbradberry committed:
Do an immediate reload upon change to the supervisor conf file

jbradberry committed:
Add the ssh pipeline option to the tasks that become a non-privileged user

Newer versions of Ansible close a security loophole by not allowing
the uploaded script snippets to be world-readable, but this causes a
problem when they need to run as an unprivileged user.

jbradberry committed:
Explicitly enable supervisord

systemd seems to have a slight problem with our pip-installed
supervisord, and needs an explict 'enable' command to run a script to
process the non-standard setup.

jbradberry committed:
Adjust the envfile to match the expectations of the settings/deploy.py

jbradberry committed:
Add support to set the number of gunicorn threads

jbradberry committed:
Add SOLR_CORE to the possible environment variables

jbradberry committed:
Support a configure option for globally installing npm packages

with sufficient flexibility that 'less' or 'sass' or whatever can be chosen.

jbradberry committed:
Fix the upload github key task condition to deal with a lack of value

jbradberry committed:
Update the readme to show that github_deploy_key is not always required

jbradberry committed:
Allow for the definition of extra environment variables as extra_env

jbradberry committed:
Change the supervisor management tasks to handlers

and trigger them appropriately over a wider set of changes,
e.g. updating the project codebase.

jbradberry committed:
Restart supervisord and all apps when code changes

`supervisorctl update` only works on conf changes, not code changes.

jbradberry committed:
Remove the handler to reread conf files from code-changing tasks

jbradberry committed:
Restart the program group when code or supervisord config changes

The 'supervisorctl' Ansible module with state=restarted triggers both
an 'update' and a 'restart'.  Update implies both a reload of the
supervisord configuration and adding (and removing) programs to the
process groups.  This should cover everything we need.

jbradberry committed:
Make the restart of supervisord managed processes unconditional

The process groups do not work the way I assumed in the previous
commit, and certain tasks that were set up to trigger the handlers
come up as 'changed' even on a trivial execution, so I feel it is
better to do this the simple way and just always restart on deploying,
rather than do something tricky that will wind up being fragile.

jbradberry committed:
Restart supervisord when the config file has changed

Do it earlier than the handler phase, since we get socket errors otherwise.

jbradberry committed:
Pin the install of gunicorn to the value of variable gunicorn_version

if defined.

jbradberry committed:
Remove the Deadsnakes PPA

it is no longer being maintained.

jbradberry committed:
Merge pull request #3 from caktus/remove-deadsnakes-ppa

Remove the Deadsnakes PPA

jbradberry committed:
Remove the optional syncdb command

It was deprecated in Django 1.7, and removed entirely in Django 1.9.
Given the age, it seems unlikely that we would upgrade a project to
Tequila but not also upgrade the version of Django.

jbradberry committed:
Move the restart gunicorn step to after the db migration

To minimize the chance that new code will attempt to access fields
that aren't there yet.

vkurup committed:
Add env vars for DB_USER and DB_NAME

trawick commented on tasks/main.yml:
```diff
@@ -80,7 +78,7 @@
-  when: github_deploy_key is defined
+  when: github_deploy_key|default('')
 
 # TODO: come up with a safer way of exposing the local repo
```
Isn't this commented-out rsync code implemented down below for task "sync source from local directory"?  IOW, it looks to me that this commented out stuff starting with "# TODO" can be removed.  Also, I'd remove "#   set_remote_user: true" in the actual rsync code.

trawick commented:
See my comment above about commented-out rsync/synchronized code.  Other than the last commit (100f87e), this matches what we've been using successfully for months on another project.  :+1: after doing the right thing with the commented-out code.

jbradberry committed:
Remove commented out code
